samcli/lib/intrinsic_resolver/intrinsic_property_resolver.py

```diff
@@ -0,0 +1,743 @@
+    verify_number_arguments, verify_intrinsic_type_str, verify_intrinsic_type_dict, verify_intrinsic_type_bool, \
+    verify_all_list_intrinsic_type, InvalidSymbolException
+
+LOG = logging.getLogger(__name__)
+
+
+class IntrinsicResolver(object):
+    AWS_INCLUDE = "AWS::Include"
+    SUPPORTED_MACRO_TRANSFORMATIONS = [AWS_INCLUDE]
+    _PSEUDO_REGEX = r'AWS::.*?'
+    _ATTRIBUTE_REGEX = r"[a-zA-Z0-9]*?\.?[a-zA-Z0-9]*?"
+    _REGEX_SUB_FUNCTION = r'\$\{(' + _PSEUDO_REGEX + '||' + _ATTRIBUTE_REGEX + r')\}'
+
+    FN_JOIN = "Fn::Join"
+    FN_SPLIT = "Fn::Split"
+    FN_SUB = "Fn::Sub"
+    FN_SELECT = "Fn::Select"
+    FN_BASE64 = "Fn::Base64"
+    FN_FIND_IN_MAP = "Fn::FindInMap"
+    FN_TRANSFORM = "Fn::Transform"
+    FN_GET_AZS = "Fn::GetAZs"
+    REF = "Ref"
+    FN_GET_ATT = "Fn::GetAtt"
+    FN_IMPORT_VALUE = "Fn::ImportValue"
+
+    SUPPORTED_INTRINSIC_FUNCTIONS = [
+        FN_JOIN,
+        FN_SPLIT,
+        FN_SUB,
+        FN_SELECT,
+        FN_BASE64,
+        FN_FIND_IN_MAP,
+        FN_TRANSFORM,
+        FN_GET_AZS,
+        REF,
+        FN_GET_ATT,
+        FN_IMPORT_VALUE
+    ]
+
+    FN_AND = "Fn::And"
+    FN_OR = "Fn::Or"
+    FN_IF = "Fn::If"
+    FN_EQUALS = "Fn::Equals"
+    FN_NOT = "Fn::Not"
+
+    CONDITIONAL_FUNCTIONS = [
+        FN_AND,
+        FN_OR,
+        FN_IF,
+        FN_EQUALS,
+        FN_NOT
+    ]
+
+    def __init__(self, symbol_resolver, template=None):
+        """
+        Initializes the Intrinsic Property class. Customers can pass in their own ref_resolver or attribute_resolver
+        to customize the behavior of ref resolution. This can also be done by extending the definition.
+
+        When created the class can either process a template or the individual attributes. The more information given
+        the easier resolution will be.
+
+        In the future, for items like Fn::ImportValue multiple templates can be provided
+        into the function.
+        Parameters
+        -----------
+        symbol_resolver: IntrinsicsSymbolTable
+            This is a required property that is the crux of the intrinsics resolver. It acts as a mapping between
+            resource_type and object to the Ref, Arn, etc.
+        template: dict
+            A dictionary containing all the resources and the attributes of the CloudFormation template. The class does
+            not mutate the template.
+        """
+        self._template = template or {}
+        self._resources = self._template.get("Resources", {})
+        self._mapping = self._template.get("Mappings", {})
+        self._parameters = self._template.get("Parameters", {})
+        self._conditions = self._template.get("Conditions", {})
+
+        self._symbol_resolver = symbol_resolver
+        self._symbol_resolver.resources = self._resources
+        self._symbol_resolver.parameters = self._parameters
+
+        self.intrinsic_key_function_map = {
```

Comment:
wouldn't a class like the following handles extending it:
```
class NewResolver(IntrinsicResolver):
    def __init__(self):
         super(IntrinsicResolver, self).__init__()
         self.intrinsic_key_function_map.extend({dic of resolvers})
```
 I'm on the fence about injecting it as a class. It feels unnecessary/unpythonic to have a class just for the dictionary rather than have it extended.

Reply:
I cleaned up the code to expose intrinsic_key_function_map to a user. I also exposed some of the properties such as adding the template when resolving it.